github/githubapp_test.go

```diff
@@ -43,52 +43,70 @@ func TestGenerateInstallationToken(t *testing.T) {
 	ts := httptest.NewServer(http.HandlerFunc(handler))
 	defer ts.Close()
 
+	app := github.GitHubApp{
+		ClientId:       clientID,
+		InstallationId: installationID,
+		PrivateKey:     string(testhelp.EncodePrivateKeyToPEM(privateKey)),
+	}
+	err = app.Validate()
+	assert.NilError(t, err)
+
 	gotToken, err := github.GenerateInstallationToken(
 		ctx,
 		ts.Client(),
 		ts.URL,
-		github.GitHubApp{
-			ClientId:       clientID,
-			InstallationId: installationID,
-			PrivateKey:     string(testhelp.EncodePrivateKeyToPEM(privateKey)),
-		},
+		app,
 	)
 
 	assert.NilError(t, err)
 	assert.Equal(t, "dummy_installation_token", gotToken)
 }
 
-func TestGitHubAppIsZero(t *testing.T) {
+func TestGitHubAppValidateSuccess(t *testing.T) {
+	privateKey, err := testhelp.GeneratePrivateKey(t, 2048)
+	assert.NilError(t, err)
+
+	app := github.GitHubApp{
+		ClientId:       "client-id",
+		InstallationId: 12345,
+		PrivateKey:     string(testhelp.EncodePrivateKeyToPEM(privateKey)),
```

Comment:
I am looking at lines 66 and 72. In both cases we own the functions that manipulate the key data (they are 2 test helpers).

What is not clear to me is this: Why are we doing 2 transformations? Why not using only 1 test helper, that produces directly a private key encoded to PEM ?

Reply:
Ah the reason for the split was I need both the `*rsa.PrivateKey` and PEM string...

e.g. in `DecodeJWT(... key *rsa.PrivateKey))` function I have to use: `&key.PublicKey`

Reply:
GitHub UI is killing me. Going back to the thread in question:

I still do not understand. Where is `privateKey` used in the test?

Reply:
> Where is privateKey used in the test?

You mean where is:
```
privateKey, err := testhelp.GeneratePrivateKey(t, 2048)
```
used?

Its used here:
```
app := github.GitHubApp{
		...
		PrivateKey:     string(testhelp.EncodePrivateKeyToPEM(privateKey)),
```
where I pass it as a parameter to function: `testhelp.EncodePrivateKeyToPEM`

Or I don't understand your question....

Reply:
So it seems that my original question is still valid:

> What is not clear to me is this: Why are we doing 2 transformations? Why not using only 1 test helper, that produces directly a private key encoded to PEM ?

?

Reply:
I thought I answered this already just above...

At first, I fell in the same trap and went to adjust according to your comment... Only to found the real reason: There are more tests... In some other tests, I need a public part of the private key in some third function: `DecodeJWT(... key *rsa.PrivateKey))` where I have to use: `&key.PublicKey`

So:
1. I create a private key with:
`privateKey, err := testhelp.GeneratePrivateKey(t, 2048)`
2. In some tests its enough to get the PEM string: `testhelp.EncodePrivateKeyToPEM`
3. in other tests I need both the PEM string and public key part of the private key OBJECT (`&key.PublicKey`)...